src/eins-timer.c

```diff
@@ -0,0 +1,210 @@
+ * checked.
+ *
+ * @error will be set to a #GIOError if, for example, the process runs out
+ * of file descriptors.
+ *
+ * Returns: (transfer full): a new `CLOCK_BOOTTIME` #GSource, or %NULL with
+ *   @error set
+ */
+static GSource *
+eins_boottime_source_new_seconds (guint    interval,
+                                  GError **error)
+{
+  g_autoptr(GSource) source = NULL;
+  EinsBoottimeSource *self = NULL;
+  int fd;
+  struct itimerspec its = {
+    .it_interval = {
+      .tv_sec = interval,
+      .tv_nsec = 0,
+    }
+  };
+
+  its.it_value = its.it_interval;
+
+  /* Set the GError if timerfd_create() fails because it could be e.g. ENFILE
+   * which we should handle gracefully */
+  fd = timerfd_create (CLOCK_BOOTTIME, TFD_CLOEXEC | TFD_NONBLOCK);
+  if (G_UNLIKELY (fd < 0))
+    {
+      g_set_error (error, G_IO_ERROR, g_io_error_from_errno (errno),
+                   "timerfd_create (CLOCK_BOOTTIME, TFD_CLOEXEC | TFD_NONBLOCK) failed: %s",
+                   g_strerror (errno));
+      return NULL;
+    }
+
+  /* But use g_error() if timerfd_settime() fails which likely would indicate
+   * programmer error */
+  /* FIXME: If this is upstreamed to GLib, handle EINVAL gracefully */
+  if (G_UNLIKELY (timerfd_settime (fd,
+                                   0,
+                                   &its,
+                                   NULL /* old_value */) < 0))
+    g_error ("timerfd_settime() failed: %s",
+             g_strerror (errno));
+
+  source = g_source_new ((GSourceFuncs *)&eins_boottime_source_funcs,
+                         sizeof (EinsBoottimeSource));
+  self = (EinsBoottimeSource *) source;
+  self->fd = fd;
+  self->tag = g_source_add_unix_fd (source, fd,
+                                    G_IO_IN | G_IO_HUP | G_IO_ERR | G_IO_NVAL);
+
+  return g_steal_pointer (&source);
+}
+
+/**
+ * eins_boottimeout_add_seconds:
+ * @interval: the timeout interval, in second
+ * @function: function to call
+ * @data: data to pass to @function
+ *
+ * Like g_timeout_add_seconds(), but uses `CLOCK_BOOTTIME` to account for time
+ * when the system is suspended.
+ *
+ * If @interval is set to zero, the #GSource will be ready the next time it's
+ * checked.
+ *
+ * Returns: the ID (greater than 0) of the event source.
+ */
+guint
+eins_boottimeout_add_seconds (guint       interval,
+                              GSourceFunc function,
+                              gpointer    data)
+{
+  guint id;
+  g_autoptr(GError) error = NULL;
+  GSource *source = eins_boottime_source_new_seconds (interval, &error);
+
+  if (!source)
+    {
+      g_error ("Failed to have EINS boottime source: %s", error->message);
+      return 0;
+    }
+
+  g_source_set_callback (source, function, data, NULL);
+  id = g_source_attach (source, NULL);
+  g_source_unref (source);
+
+  return id;
+}
+
+/**
+ * eins_get_realtime:
+ *
+ * Like g_get_real_time(), but the precision is only seconds.
```

Comment:
This function returns `guint` which is a 4-byte type on x86-64. `struct timespec`'s `tv_sec` field is a `time_t` which is 8 bytes wide. https://en.wikipedia.org/wiki/Year_2038_problem .

 `g_get_real_time()` returns `gint64`. I suggest you just use `g_get_real_time()` at each call site, and either divide it by `G_USEC_PER_SEC` or just work in microseconds. I don't think there's much advantage of aligning to seconds.

(Using `g_timeout_add_seconds` rather than `g_timeout_add` allows GLib to merge nearby timers together to avoid unnecessary wakeups, but that is not relevant here – we are using a kernel-provided timerfd, and I can't see any way to tell the kernel "it doesn't need to be precise".)

Reply:
It is due to the interval's type guint of `g_timeout_add_seconds()`.  Working in microseconds sounds like a good idea.